src/ddsrt/src/event.c

```diff
@@ -0,0 +1,384 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include "dds/ddsrt/event.h"
+#include "dds/ddsrt/event_pipe.h"
+
+/**
+* ddsrt_event creation function
+*
+* mon_type: type of object being monitored
+* ptr_to_mon: pointer to object being monitored
+* mon_sz: size in bytes of object being monitored
+* evt_type: type of event being monitored for
+*/
+struct ddsrt_event* ddsrt_event_create(enum MONITORABLE_TYPE mon_type, void* ptr_to_mon, unsigned int mon_sz, int evt_type) {
+    if (0 == mon_sz ||
+        sizeof(void*) < mon_sz) return NULL;
+    if (NULL == ptr_to_mon &&
+        MONITORABLE_UNSET != mon_type) return NULL;
+    struct ddsrt_event* returnptr = malloc(sizeof(struct ddsrt_event));
+    if (NULL == returnptr) return NULL;
+
+    void* tmpptr = malloc(8);
+    if (NULL == tmpptr) {
+        free(returnptr);
+        return NULL;
+    }
+
+    if (NULL != ptr_to_mon) memcpy(tmpptr, ptr_to_mon, mon_sz);
+    else memset(tmpptr, 0x0, mon_sz);
+
+    returnptr->mon_type = mon_type;
+    returnptr->evt_type = evt_type;
+    returnptr->mon_ptr = tmpptr;
+    returnptr->mon_sz = mon_sz;
+
+    return returnptr;
+}
+
+/**
+* ddsrt_event copy function
+*
+* dst: destination to copy to
+* src: source to copy from
+*/
+void ddsrt_event_copy(struct ddsrt_event* dst, struct ddsrt_event* src) {
+    dst->mon_type = src->mon_type;
+    dst->evt_type = src->evt_type;
+    if (dst->mon_sz != src->mon_sz) {
+        free(dst->mon_ptr);
+        dst->mon_sz = src->mon_sz;
+        dst->mon_ptr = malloc(dst->mon_sz);
+    }
+    memcpy(dst->mon_ptr, src->mon_ptr, dst->mon_sz);
+}
+
+/**
+* ddsrt_event cleanup function
+*
+* evt: event to clean up
+*/
+void ddsrt_event_destroy(struct ddsrt_event* evt) {
+    free(evt->mon_ptr);
+    free(evt);
+}
+
+/**
+* ddsrt_event container struct
+*
+* Stores the events in an array of pointers, the memory for the events is pre-assigned.
+* Adding events will make use of the pre-assigned events.
+* If the container can be expanded, adding events will cause a doubling of the size when nevents == cevents.
+*
+* events: array of pointers to stored events
+* nevents: currently stored number of events
+* cevents: current capacity for events
+* fixedsize: whether the container can be expanded after creation
+*/
+struct event_container {
+    struct ddsrt_event** events;   /*events container*/
+    int                 nevents;    /*number of events stored*/
+    int                 cevents;    /*capacity for events at this moment*/
+    int                 fixedsize;  /*whether the size of the container can be modified*/
+    //add mutex?
```

Comment:
Don't know. Locking doesn't have to be done per-object as is the case in `ddsc` in my opinion, so I'm inclined to say no. If necessary, we can do so in the future and this comment can be removed.